rnd/autogpt_server/autogpt_server/executor/manager.py

```diff
@@ -402,62 +396,89 @@ class Executor:
     @classmethod
     def on_node_executor_start(cls):
         configure_logging()
-        cls.logger = logging.getLogger("node_executor")
+
         cls.loop = asyncio.new_event_loop()
+        cls.pid = os.getpid()
+
         cls.loop.run_until_complete(db.connect())
         cls.agent_server_client = get_agent_server_client()
 
+        # Set up shutdown handlers
+        cls.shutdown_lock = threading.Lock()
+        atexit.register(cls.on_node_executor_stop)  # handle regular shutdown
+        signal.signal(  # handle termination
+            signal.SIGTERM, lambda _, __: cls.on_node_executor_sigterm()
+        )
+
+    @classmethod
+    def on_node_executor_stop(cls):
+        if not cls.shutdown_lock.acquire(blocking=False):
+            return  # already shutting down
+
+        logger.info(f"[on_node_executor_stop {cls.pid}] ⏳ Disconnecting DB...")
+        cls.loop.run_until_complete(db.disconnect())
+        logger.info(f"[on_node_executor_stop {cls.pid}] ✅ Finished cleanup")
+
+    @classmethod
+    def on_node_executor_sigterm(cls):
+        llprint(f"[on_node_executor_sigterm {cls.pid}] ⚠️ SIGTERM received")
+        if not cls.shutdown_lock.acquire(blocking=False):
+            return  # already shutting down, no need to self-terminate
+
+        llprint(f"[on_node_executor_sigterm {cls.pid}] ⏳ Disconnecting DB...")
+        cls.loop.run_until_complete(db.disconnect())
```

Comment:
Isn't this actually fine to be called twice? why mutex is needed ?

Reply:
I'm not going to make assumptions about how `db.disconnect()` works. I just don't want cleanup handlers to mess with each other. If `on_node_executor_stop` is running, and the process gets a `SIGTERM`, `on_node_executor_sigterm` will start while the executor and its resources are in an undefined/unknown state. And as long as the signal handler is active, the main thread is paused so that undefined state is possibly never resolved.

Reply:
I see so it's actually different thread? the main thread and the signal handler ? that's why the threading.lock is needed?

Reply:
I don't think it's technically a separate thread, but it behaves similarly. According to the docs, the main thread is paused and the signal handler is executed in the context of the main thread.